dask/dataframe/core.py

```diff
@@ -4746,3 +4734,26 @@ def partitionwise_graph(func, name, *args, **kwargs):
         else:
             pairs.extend([arg, None])
     return blockwise(func, name, 'i', *pairs, numblocks=numblocks, concatenate=True, **kwargs)
+
+
+def meta_warning(df):
+    """
+    Provide an informative message when the user is asked to provide metadata
+    """
+    if is_dataframe_like(df):
+        meta_str = {k: str(v) for k, v in df.dtypes.to_dict().items()}
+    elif is_series_like(df):
+        meta_str = (df.name, str(df.dtype))
+    else:
+        meta_str = None
+    msg = ("\nYou did not provide metadata, so Dask is running your "
+           "function on a small dataset to guess output types. "
+           "It is possible that Dask will guess incorrectly.\n"
+           "To provide an explicit output types or to silence this message, "
+           "please provide the `meta=` keyword, as described in the map or "
+           "apply function that you are using..")
+    if meta_str:
+        msg += ("\n"
+                "  Before: .apply(func)\n"
+                "  After:  .apply(func, meta=%s)\n" % str(meta_str))
```

Comment:
Do we also know the method they are using, or is that trying to go too far here?

Reply:
I looked into fixing this but it looks like we only use this message on apply currently.

Reply:
OK then - definitely an improvement.